### ISSUE TITLE
Cascades should participate in null shortening.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#1171]
Currently cascades do not participate in null shortening.
If you have:
```dart
class C {
  List<String> buffer = ...
}
```
you can write:
```
  C c = ...;
  c.buffer.add("0");
  c.buffer..add("a")..add("b");
```
If `c` is nullable, and you want to add to it if non-null, you currently have to write:
```dart
  C? c = ...;
  c?.buffer.add("0");
  c?.buffer?..add("a")..add("b");
```
I would expect `c?.buffer..add("a")..add("b");` to work, but it doesn't because the null shortening stops before the cascade.

All the same reasons we used for making `?.` null shortening applies equally well to the cascades.
The `?..` should be unnecessary, and the `?.` should null-shorten the entire cascade as well.

**Comment by leafpetersen** [comment]
What about:

```dart
extension Show on Object? {
  void show() => print(this.toString());
}

void main() {
  Map<int, String> l = {0: "hello", 2: "world"};
  l..[0]?.show()..[1]?.show()..[2]?.show();
}
```

Are you proposing that this would only print `"hello"`, or would it still print `"hello\nworld"`?  The former feels very unexpected to me.

**Comment by lrhn** [comment]
It should print `"hello\nworld\n"`.

I don't propose to change associativity, only the extent of null-shortening.

Currently `x?.y..z..w` will complain that you cannot do `..z` on a nullable value. The null shortening ends before the `..`. I want the null shortening to include the `..z` and `..w` cascade sections, just as it would a single `.z` or `.w` selector.

One possible mental model for null shortening is that `x?.selectors` evaluates the same as `x.selectors` except that:
* When `x` is `null`, it is instead replaced by a special "propagating null" value.
* Any selector (which I want to include cascade selectors) evaluated on a propagating null is a no-op which again evaluates to a propagating null
* At some point, at the end of the null-shortening, we convert a propagating null back to a normal `null`.  (And what we've been discussing is where this happens).

Here `l..[0]?.show()..[1]?.show()..[2]?.show()` will evaluate `l` to a value
* Read `[0]` from `l`.
* See that that value is not `null`.
* Invoke `.show()` on the value.
* selector chain ends here, so if the result was a propagating null, it would become a normal null. It isn't, so it won't.
* Then the result is ignored because the cascade section ended.
* Starting the next cascade section, we read `[1]` from `l`
* Then we detect that the value *is* `null`, and convert it to a propagating null.
* Then the `.show()` invocation does nothing because it's called on a propagating null.
* Then the propagating null is converted back to a normal `null`.
* Then the value is ignored.
* Starting the next cascade section
* ... same as the first one.


For the example above, `x?.y..z..w`, evaluation would do;
* Evaluate `x` to a value
* Let's assume that value is `null`, so it is replaced by a propagating `null`.
* Invoking `.y` on a propagating `null` does nothing.
* Currently we convert back to a normal `null` here. If we didn't, we'd continue as:
* Do nothing for `..z`.
* Do nothing for `..w`.
* We are done, so convert back no normal `null`.

So, when I say that cascades should participate in null shortening, I mean that we should not convert a propagating null back to a normal null when seeing a cascade selector, `..selector`. We should instead let null shortening skip the cascade section(s), in the same way we would if we'd had a null aware cascade like `(null as Y)?..z..w`.

That is 
```dart
var xy = x?.y; 
xy?..z..w
``` 
and 
```dart
x?.y..z..w
``` 
will both skip `.y`, `.z` and `.w`, and we don't need the intermediate `?..` when we keep it as one expression. Just like `x?.y.z` doesn't need an extra `?.` if we don't split it as `var xy = x?.y; xy?.z;`.

(This is all very coherent in my head, I hope I make some kind of sense! :sad:)

**Comment by leafpetersen** [comment]
I think I'm on board with this.  @munificent @eernstg @jakemac53 @natebosch @stereotype441 any concerns?  The proposal here is that null shorting in the *target* of a cascade should propagate past the cascade itself (this has nothing to do with null shorting within a cascade section, which remains contained to that section).  See my example a couple of comments up, and the reply from @lrhn .

**Comment by stereotype441** [comment]
I like the idea of making cascades participate in null shortening.  It's easy to explain, it's useful (though perhaps marginally so), and I think it's a better match for user expectations than the behavior we currently have.

I have some vague concerns about feasibility.  The logic in the analyzer to support cascades is pretty hairy, and so is the logic for supporting null shorting.  Put the two together and, well, hopefully everything will work out just fine!  But there's some risk that we'll need to do some ugly refactor.

An example of the hairiness is that whereas the grammar for cascades in the language spec looks like this:
```
<cascade> ::= <cascade> `..` <cascadeSection>
           |  <conditionalExpression> (`?..` | `..`) <cascadeSection>
```
the analyzer's AST representation is structured more like:
```
<cascade> ::= <conditionalExpression> ((`?..` | `..`) <cascadeSection>)+
```
(that is, the analyzer's `CascadeExpression` object stores a single target expression and a list of cascade sections).  I _think_ it'll be ok, but it's a source of scariness because if we *do* have to change it, we'll have to be pretty deliberate about it, since the analyzer's AST representation is exposed through its public API.

My biggest concern, though, is about schedule impact.  We've been doing a lot of polishing of subtle corners of null safety recently (number promotion, flow analysis tweaks), and at some point we're going to have to stop polishing and accept some warts so that we can ship the feature.  My gut feeling is that we could leave this particular wart unpolished and virtually no one would notice or care 😃

**Comment by munificent** [comment]
I'm not strongly opposed, but I lean slightly against it. Like @stereotype441 notes, there are real implementation and schedule concerns.

Also, I think it may be very surprising that a `?.` in one position can cause shorting that goes *beyond* the extent of a *later* `?.`:

```dart
ape?.bat..cat.dog()..eel().fly();
// '---------------------------^
// shorts all the way to

ape.bat..cat?.dog()..eel().fly();
//          '-----^
//          only shorts to
```

I mean, cascade syntax is bananas *anyway*, so maybe this is not the most crazy part about it, but it feels dubious to me. Given how much potential confusion we already have around `?.` in general, null shorting, and cascades *independently of each other*, we may want to just stop the train and not add more cleverness when you combine them. I think the benefit is marginal and the complexity is real.

**Comment by eernstg** [comment]
Like @munificent, I'm slightly sceptical about this feature. If it can be implemented without disrupting other work too much, I can see that it might be seen as a consistent and useful extension of null shortening.

However, I think the need for this feature is smaller than the need for the null shortening that we already have. The existing null shortening allows developers to have `.` rather than `?.` in many locations where the older semantics required `?.` and conflated the distinction between the propagating null and a null result coming from the latest selector.

With cascades, we already propagate the initial null (because it's `r?..c1..c2..c3`, not `r?..c1?..c2?..c3`), so the only `?` which is eliminated is at the front. Moreover, that `?` doesn't conflate a propagating null and a null from "the latest x" for any x, because there is always exactly one receiver expression in front of `?..`. So the benefit (in terms of avoiding `?`) is always very limited.

Another reason why it may be useful to keep things as they are today is that the cascade is a "larger" construct than the selector chain: The receiver part of a cascade is a `<conditionalExpression>` where no cascade can occur unless it is parenthesized, and a `<cascadeAssignment>` similarly assigns an expression with no cascades to its target. In other words, selector chains occur all the time inside cascades, but cascades never occur inside selector chains (unless passed as arguments or otherwise parenthesized)

Being the "big" structure, it seems reasonable to me that the cascade gets an explicit null shortening marker. For example:

```dart
void main() {
  42.isEven.someMethod(true).anotherMethod('hest')?.substring(1)
    ?..length.isEven
    ..contains('z');
}
```

In this example `?..` clearly indicates that the receiver may be null and the cascade sections will be skipped. If we do a lot of work in order to be able to omit that `?` then the developer who is reading this code will need to spot the `?` in the previous line, which is bad for the code readability. Note that we _do_ require the`?..` when the receiver is nullable for a reason which is not `?.`.

**Comment by lrhn** [comment]
>  The receiver part of a cascade is a `<conditionalExpression>`

That's actually really surprising. It means that `x ? y : z..foo` parses as `(x ? y : z)..foo`. 

I'd like to fix that, probably by doing something like changing:
```
{\bf expression:} ...
       conditionalExpression cascadeSections?;
```
to
```
{\bf expression:} ...
 conditionalExpression;
 primaray selector* cascadeSections;
```
And then the cascade *will* fit into the selector chain.

**Comment by eernstg** [comment]
That could be a massively breaking change: The current grammar allows almost all expressions to be the receiver in a cascade (from `e1 ?? e2..m()` down to `await e..m()` and `x++..bar()`).

**Comment by lrhn** [comment]
Damn. I think we might have been around that before. :cry:
So, that'll have to be another *breaking* change where we force people to parenthesize their non-selector-chain cascade receivers.
(I still want it, though!)

**Comment by eernstg** [comment]
> we might have been around that before

Sure. :smiley:

I think the standard formatting of cascades (and the fact that they tend to be rather large) would typically put each `?..`/`..` at the beginning of a line. That makes it stand out visually as part of a bigger structure than each element of a receiver (in particular, each selector in a receiver which is a selector chain).

**Comment by lrhn** [comment]
Couldn't we still let cascades participate in null shortening when they are following a selector chain (which is the only case where they are in scope of a null-shortening operator at all).

**Comment by eernstg** [comment]
Right, but that actually brings up another source of confusion:

```dart
int? i = null;
1 + i?.toString().length..isEven;
```

The `?.` will remove the need for `?..` in `i?.toString().length..isEven`, but with `1 +` in front we will just get an error because `+` doesn't accept a nullable argument.

**Comment by lrhn** [comment]
Writing `1+ i?.toString().length..isEven` would associate as `(1+i?.toString(0.length)..isEven`. The `..isEven` is not continuing a selector chain, so it's defintiely not participating in the null shortening of that selector chain, and the `?.` won't remove any need for `?..`. 
The `?..` isn't needed because `1+` returns `num`, and then it's an error that the second operand is `int?`.

It's completely bonkers to associate it like that, and why I'd want to change it so that it actually associates as `1 + (i?.toSTring().length..isEven)`. It'll still be an error that the second operand is `int?`.

**Comment by eernstg** [comment]
Well that was the point: We have a `?.` selector chain which is immediately followed by a cascade, and it is still not true that the null shorting includes the cascade.

This again illustrates that the cascade is a "larger" construct than the selector chain, and hence it's probably not helpful to try to make the cascade a part of the selector chain in terms of null shorting or other non-syntactic perspectives.

**Comment by lrhn** [comment]
I agree that it's hard to parse because `1 + x?.y..z` makes it look like the `..z` which *follows* `x?.y` is also associated with it.
That's independent of null shortening, it's completely unreadable to begin with.

I'm just saying that when a cascade *does* continue a selector chain (which it doesn't when it continues an a addition, no matter how the source looks), then I want the cascade to participate in null-shortening of that selector chain.
I want that because *it's useful*, there is a way out when you don't want it (parentheses), it's actually what users will expect. 
The fact that cascades bind oddly won't matter much in practice because nobody will be writing `1 + something.something..foo` anyway. They'll quickly figure out that it doesn't do what they want and rewrite it with parentheses anyway. I dare you to find a real-world cascade which isn't on a primary, object creation or selector.

**Comment by eernstg** [comment]
I agree that `1 + x?.y..z` is hard to read in the first place. We do have cases like [this](https://github.com/dart-lang/sdk/blob/62093ca8ede6cac91c9724907b31a05bb85cac56/runtime/observatory/lib/src/elements/class_tree.dart#L132):

```dart
    return (await Future.wait(cls.subclasses.map(_getActualChildrens)))
        .expand((f) => f)
          ..forEach((subcls) {
            _mixins[subcls.id] = (_mixins[subcls.id] ?? [])..add(cls.mixin);
          });
```

where the author _did_ include the parentheses around `_mixins[subcls.id] ?? []`, but could have omitted them. So this confirms that the "hard to read" factor can make developers add the parentheses in any case.

But I still think that the motivation for omitting `?` in `?..` is relatively weak: It's already non-repeated, and the formatting where `..` is placed at the beginning of a line makes `?..` helpful. Otherwise we can have

```dart
some(2).longish?.thingy['xyz'].that.contains(42).a.hidden().questionMark
    ?..andThenALongishBunchOfCascadeSections
    ..andThenALongishBunchOfCascadeSections
    ..andThenALongishBunchOfCascadeSections;
```

and it's really not an improvement to change `?..` to `..`.

About the grammar change: I agree that the derivation rule in https://github.com/dart-lang/language/issues/1171#issuecomment-680786569 allows for `<selectorChain>`s in the same way as today, and everything else can be parenthesized (because `( <expression> )` is a `<primary>`), but we might want to fiddle a bit more with it, e.g., to allow `await e..cascadeSection`.

**Comment by leafpetersen** [comment]
I don't think we have consensus to move forward on this, any strong objections to closing this out?

**Comment by lrhn** [comment]
Let's close for now. If we go back to looking at the precedence of cascades at a later point, we may want to include null-shortening as well, but we'll probably not do it by itself.
